tools/workqlat.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/libexec/platform-python
+# @lint-avoid-python-3-compatibility-imports
+#
+# workqlat   Work queue latency as a histogram.
+#           For Linux, uses BCC, eBPF.
+#
+# USAGE: workqlat [-w] [interval] [count]
+#
+# This measures the time a work item spends waiting on a work queue before
+# being executed by a worker, and shows this time as a histogram. This time
+# should be small, but execution of work item may get delayed due to CPU load.
+#
+#
+# Copyright 2022 Oracle and/or its affiliates.
+# Licensed under the Apache License, Version 2.0 (the "License")
+#
+# Inspired by runqlat
+# 06-Jan-2022   Imran Khan   Created this.
+
+from __future__ import (
+    absolute_import, division, unicode_literals, print_function
+)
+from bcc import BPF
+from time import sleep
+import argparse
+
+# arguments
+examples = """examples:
+    ./workqlat           # summarize work queue latency as a histogram
+    ./workqlat 1 10      # print 1 second summaries, 10 times
+    ./workqlat -W        # show each work-handler item separately
+    ./workqlat -w xyz    # show specified work handler
+"""
+parser = argparse.ArgumentParser(
+    description="Summarize work queue latency as a histogram",
+    formatter_class=argparse.RawDescriptionHelpFormatter,
+    epilog=examples)
+parser.add_argument("-W", "--works", action="store_true",
+    help="print a histogram per work-handler item")
+parser.add_argument("-w", "--work",
+    help="trace only specified work item")
+parser.add_argument("interval", nargs="?", default=99999999,
+    help="output interval, in seconds")
+parser.add_argument("count", nargs="?", default=99999999,
+    help="number of outputs")
+parser.add_argument("--ebpf", action="store_true",
+    help=argparse.SUPPRESS)
+arguments = parser.parse_args()
+countdown = int(arguments.count)
+debug = 0
+
+src = """
+#include <uapi/linux/ptrace.h>
+
+typedef struct work_key {
+    u64 work;
+    u64 handle;
+} work_key_t;
+
+typedef struct work_handle_key {
+    u64 handle;
+    u64 slot;
+} work_handle_key_t;
+
+BPF_HASH(record, work_key_t);
+
+STORAGE
+
+TRACEPOINT_PROBE(workqueue, workqueue_queue_work) {
+    work_key_t key = {.work = 0, .handle = 0};
+
+    FILTER
+
+    u64 ts = bpf_ktime_get_ns();
+    if (args != NULL) {
```

Comment:
Why do you check `args` against `NULL` ?

Reply:
This was wrong because tracing framework would not allow a NULL args. I got overcautious there.